lib/plugins/aws/deploy/lib/checkForChanges.js

```diff
@@ -289,20 +290,55 @@ module.exports = {
     const subscriptionFilterComparator = (sf1, sf2) =>
       sameDestinationArn(sf1, sf2) && sameLogicalId(sf1, sf2);
 
-    return Promise.all(
-      oldSubscriptionFilters.map((oldSubscriptionFilter) => {
+    const matchedOldSubscriptionFilters = oldSubscriptionFilters.filter((oldSubscriptionFilter) => {
+      const destinationMatchNewSubscriptionFilter = newSubscriptionFilters.find(
+        (newSubscriptionFilter) => sameDestinationArn(newSubscriptionFilter, oldSubscriptionFilter)
+      );
+      return destinationMatchNewSubscriptionFilter;
+    });
+
+    const notMatchedOldSubscriptionFilters = oldSubscriptionFilters.filter(
+      (oldSubscriptionFilter) => {
+        const destinationMatchNewSubscriptionFilter = newSubscriptionFilters.find(
+          (newSubscriptionFilter) =>
+            sameDestinationArn(newSubscriptionFilter, oldSubscriptionFilter)
+        );
+        return !destinationMatchNewSubscriptionFilter;
+      }
+    );
+
+    const exceedSubscriptionFilterCount =
+      notMatchedOldSubscriptionFilters.length +
+      newSubscriptionFilters.length -
+      CLOUDWATCHLOG_LOG_GROUP_EVENT_PER_FUNCTION_LIMIT;
+    if (exceedSubscriptionFilterCount > 0) {
+      await Promise.all(
+        notMatchedOldSubscriptionFilters
+          .slice(0, exceedSubscriptionFilterCount)
+          .map((oldSubscriptionFilter) =>
+            this.provider.request('CloudWatchLogs', 'deleteSubscriptionFilter', {
+              logGroupName,
+              filterName: oldSubscriptionFilter.filterName,
+            })
+          )
+      );
+    }
+
+    const orderChangedOldSubscriptionFilters = matchedOldSubscriptionFilters.filter(
```

Comment:
code looks messy, in following work, it should be like
1. exceed limit by other stack's subscriptions filter -> throw error at check deploy state
2. delete all not-matched subscription filter in current stack at check deploy state(include both order changed or deleted ones)
3. exceed limit by current stack -> it will throw error at compile state

Reply:
What does `orderChangedOldSubscriptionFilters` means? Why do we have to remove them?

Reply:
not sure why.. but I found one test case which expect it to be deleted
https://github.com/serverless/serverless/blob/master/test/unit/lib/plugins/aws/deploy/lib/checkForChanges.test.js#L1252

My guess is when new CW log event added, like below scenario
first, stack is created like following
```
functions:
  fn1:
    handler: handler.hello
    events:
      - cloudwatchLog: '/test/log/hello'
  fn2:
    handler: handler.hello
    events:
      - cloudwatchLog: '/test/log/hello'
```

and decide to change like this (add other CW log event above existing one)
```
functions:
  fn1:
    handler: handler.hello
    events:
      - cloudwatchLog: '/test/log/bye'
      - cloudwatchLog: '/test/log/hello'
  fn2:
    handler: handler.hello
    events:
      - cloudwatchLog: '/test/log/hello'
```

fn1's previous subscription of 'test/log/hello' should be deleted before create the new one
cause currently, subscription filters are depends on the index of the events in function

Reply:
If its not necessary, I think we can remove it

Reply:
small addition) If we can make subscription filter logical id with CW log group name, it will be much better
But It may break all the current stacks 😅

Reply:
Oh, I see now, it looks like the reason for it is the fact that we're actually using the order in which the events are defined, so we have to maintain proper ordering, but I would assume that actual reordering is not going to happen very often here. Making subscription filter logical id with CW log group name could be great, but as you mention, it would potentially be a breaking change for existing stack, so we might consider it as a future improvement behind a flag with deprecation warning.